src/rust/ide/src/controller/searcher.rs

```diff
@@ -578,57 +587,150 @@ impl Searcher {
+            let call     = SimpleFunctionCall::try_new(&call_ast)?;
+            if let Some(module) = self.module_method_called(&call) {
+                let entry = self.database.lookup_by_name_and_module(call.function_name,&module);
+                Some(entry.into_iter().collect())
+            } else {
+                let name     = &call.function_name;
+                let location = *self.position_in_code;
+                Some(self.database.lookup_by_name_and_location(name,location))
+            }
+        };
+        opt_result().unwrap_or(Vec::new())
+    }
+
+    fn module_method_called(&self, call:&SimpleFunctionCall) -> Option<QualifiedName> {
+        let module_ast      = self.graph.graph().module.ast();
+        let position        = *self.position_in_code;
+        let module          = double_representation::module::Info {ast:module_ast};
+        let this_name       = ast::identifier::name(call.this_argument.as_ref()?)?;
+        let matching_locals = self.database.lookup_locals_by_name_and_location(this_name,position);
+        let not_local_name  = matching_locals.is_empty();
+        let imported        = || module.iter_imports().find_map(|import| {
+            import.qualified_name().ok().filter(|module| module.name() == this_name)
+        });
+        not_local_name.and_option_from(imported)
+    }
+
+    /// Get the suggestion that was selected by the user into the function.
+    ///
+    /// This suggestion shall be used to request better suggestions from the engine.
+    fn intended_function_suggestion(&self) -> Option<CompletionSuggestion> {
+        let id       = CompletedFragmentId::Function;
+        let fragment = self.data.borrow().find_picked_fragment(id).cloned();
+        fragment.map(|f| f.picked_suggestion.clone_ref())
     }
 
     /// Returns the Id of method user intends to be called in this node.
     ///
     /// The method may be picked by user from suggestion, but there are many methods with the same
     /// name.
     fn intended_method(&self) -> Option<MethodId> {
-        let borrowed_data = self.data.borrow();
-        let mut fragments = borrowed_data.fragments_added_by_picking.iter();
-        let function_frag = fragments.find(|frag| frag.id == CompletedFragmentId::Function)?;
-        function_frag.picked_suggestion.method_id()
+        self.intended_function_suggestion()?.method_id()
+    }
+
+    /// If the function fragment has been filled and also contains the initial "this" argument.
+    ///
+    /// While we maintain chain consisting of target and applied arguments, the target itself might
+    /// need to count for a one argument, as it may of form `this.method`.
+    fn has_this_argument(&self) -> bool {
+        self.data.borrow().input.expression.as_ref()
+            .and_then(|expr| ast::opr::as_access_chain(&expr.func))
+            .is_some()
+    }
+}
+
+
+struct SimpleFunctionCall {
+    this_argument : Option<Ast>,
+    function_name : String,
+}
```

Comment:
Shouldn't this also include support for the non sugared universal invocation syntax `method this`?

Reply:
In such case the `this_argument` is None.

O, or perhaps we should rename `this_argument` to the `dot_applied_argument`

Reply:
This is a little more complicated and not in the task scope, therefore I leave it as it is now.